python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -2394,6 +2399,42 @@ def get_handlers(self) -> Sequence[logging.Handler]:
     def store_event(self, event: "EventLogEntry") -> None:
         self._event_storage.store_event(event)
 
+    def should_retry_run(self, run: DagsterRun) -> bool:
+        from dagster._core.events import RunFailureReason
+
+        if run.status == DagsterRunStatus.FAILURE:
+            raw_max_retries_tag = run.tags.get(MAX_RETRIES_TAG)
+            if raw_max_retries_tag is None:
+                max_retries = self.run_retries_max_retries
+            else:
+                try:
+                    max_retries = int(raw_max_retries_tag)
+                except ValueError:
+                    warnings.warn(
+                        f"Error parsing int from tag {MAX_RETRIES_TAG}, won't retry the run."
+                    )
+                    return False
+            if max_retries > 0:
+                run_group = self.get_run_group(run.run_id)
+                if run_group is not None:
+                    _, run_group_iter = run_group
+                    # since the original run is in the run group, we retry one more time
+                    # if the length of the run group is max_retries
+                    if len(list(run_group_iter)) <= max_retries:
```

Comment:
So trying to understand the potential for race condition here; if another run kicks off at the same time as this loop is being evaluated, there's potential for us to actually have an extra retry here right? Might be worth noting that.

Reply:
I don't think there is a race condition here. What i was trying to say in the comment is that the original run is included in the run group, so if we want to have 5 retries, that means when the run group is length 5, only 4 retries have been launched since the run group is (original_run, retry_1, retry_2, retry_3, retry_4)

Reply:
maybe the only race condition is if max_retries is 5, and we've launched 4 retries already. We get the run group `(original_run, retry_1, retry_2, retry_3, retry_4)`, and right after that someone manually launches a retry of the run. In that case, the auto-reexecution system would launch a retry (since only 4 retries had been launched), but the manual retry would also be part of the run group, making the run group larger than the number of retries expected. 


I think this is fine though:
1. this race condition exists in the current system 
2. it seems really unlikely 
3. a user can launch a manual retry whenever they want (ir long after all automatic retries have been attempted) and we don't have any restrictions on that

Reply:
Why is the case that I mentioned not possible? Definitely seems unlikely but I'm imagining two runs in the same group occurring simultaneously and this loop evaluates simultaneously for them

Reply:
Actually in hindsight I think we're talking about the same case- that essentially manual run retries can occur whenever and can make the logic in this loop out of date. I think it's worth adding a comment for that

Reply:
yeah i misunderstood your first comment. 

So your thinking is we have max retries at 5 and we have (original_run, retry_1, retry_2) then then two runs (say one automatic and one manual) get launched and fail at the same time, and this fn is executed at the same time. 

Then the run group would be (original_run, retry_1, retry_2, retry_3, manual_retry), so both retry_3 and manual_retry would be marked as `will_retry`

Reply:
that's right yea

Reply:
added the comment - lmk if the description of the race condition makes sense to you